[export] Fix state dict device serialization

angelayi opened the pull request:
Summary:
Fixes https://github.com/pytorch/pytorch/issues/114000
Will check with SherlockNoMad on why we need to convert to cpu after his PTO

Test Plan: CI

Differential Revision: D51629068




cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @suo @ydwu4

ydwu4 commented on torch/_export/serde/serialize.py:
```diff
@@ -943,12 +944,23 @@ def serialize(self, graph_module: torch.fx.GraphModule) -> GraphModule:
+    def __init__(
+        self,
+        opset_version: Optional[Dict[str, int]] = None,
+        convert_to_cpu: bool = False,
```
Do we need to store the original devices for each tensor somewhere so that we could retrieve the original devices?

ydwu4 commented on torch/_export/serde/serialize.py:
```diff
@@ -943,12 +944,23 @@ def serialize(self, graph_module: torch.fx.GraphModule) -> GraphModule:
             self.opset_version["aten"] = torch._C._get_max_operator_version()
+        self.convert_to_cpu = convert_to_cpu
 
     def serialize(self, exported_program: ep.ExportedProgram) -> SerializedArtifact:
```
compared with a self.convert_to_cpu attibute, can we make it a default kwarg to serialize? This makes the Serializer less stateful.

ydwu4 reviewed (approved):
Left a few non-blocking comments. Overall looks good to me.

angelayi committed:
[export] Fix state dict device serialization (#114695)

Summary:

Fixes https://github.com/pytorch/pytorch/issues/114000
Will check with SherlockNoMad on why we need to convert to cpu after his PTO

Test Plan: CI

Reviewed By: ydwu4

Differential Revision: D51629068

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -950,7 +951,17 @@ def __init__(self, opset_version: Optional[Dict[str, int]] = None):
+        self,
+        exported_program: ep.ExportedProgram,
+        *,
+        convert_to_cpu=False,
```
I think we should have a unified behavior on serialize API.

zhxchen17 reviewed (changes requested):
the way we add flags everywhere looks sus...
